File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureV2UiCustomizationUnitTest.java
package com.braintreepayments.api.threedsecure;

import android.os.Parcel;

import com.braintreepayments.api.threedsecure.ThreeDSecureV2ButtonCustomization;
import com.braintreepayments.api.threedsecure.ThreeDSecureV2LabelCustomization;
import com.braintreepayments.api.threedsecure.ThreeDSecureV2TextBoxCustomization;
import com.braintreepayments.api.threedsecure.ThreeDSecureV2ToolbarCustomization;
import com.braintreepayments.api.threedsecure.ThreeDSecureV2UiCustomization;
import com.cardinalcommerce.shared.models.enums.ButtonType;
import com.cardinalcommerce.shared.userinterfaces.UiCustomization;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

@RunWith(RobolectricTestRunner.class)
public class ThreeDSecureV2UiCustomizationUnitTest {

    @Test
    public void setsAllCardinalClassProperties() {
        ThreeDSecureV2UiCustomization sut = new ThreeDSecureV2UiCustomization();
        sut.setLabelCustomization(new ThreeDSecureV2LabelCustomization());
        sut.setButtonCustomization(new ThreeDSecureV2ButtonCustomization());
        sut.setButtonType(ThreeDSecureV2ButtonType.BUTTON_TYPE_NEXT);
        sut.setTextBoxCustomization(new ThreeDSecureV2TextBoxCustomization());
        sut.setToolbarCustomization(new ThreeDSecureV2ToolbarCustomization());

        UiCustomization cardinalUiCustomization = sut.getCardinalUiCustomization();
        assertNotNull(cardinalUiCustomization.getLabelCustomization());
        assertNotNull(cardinalUiCustomization.getButtonCustomization(ButtonType.NEXT));
        assertNotNull(cardinalUiCustomization.getTextBoxCustomization());
        assertNotNull(cardinalUiCustomization.getToolbarCustomization());
    }

    @Test
    public void constructor_defaultsUiCustomizationPropertyToEmptyObject() {
        ThreeDSecureV2UiCustomization sut = new ThreeDSecureV2UiCustomization();
        assertNotNull(sut.getCardinalUiCustomization());
    }

    @Test
    public void writeToParcel() {
        ThreeDSecureV2ButtonCustomization buttonCustomization =
                new ThreeDSecureV2ButtonCustomization();
        buttonCustomization.setBackgroundColor("#FF0000");
        buttonCustomization.setCornerRadius(5);
        buttonCustomization.setTextColor("#000000");
        buttonCustomization.setTextFontName("Comic Sans");
        buttonCustomization.setTextFontSize(20);

        ThreeDSecureV2LabelCustomization labelCustomization =
                new ThreeDSecureV2LabelCustomization();
        labelCustomization.setHeadingTextColor("#FFFFFF");
        labelCustomization.setHeadingTextFontName("Times New Roman");
        labelCustomization.setHeadingTextFontSize(30);
        labelCustomization.setTextColor("#121212");
        labelCustomization.setTextFontName("Helvetica");
        labelCustomization.setTextFontSize(15);

        ThreeDSecureV2TextBoxCustomization textBoxCustomization =
                new ThreeDSecureV2TextBoxCustomization();
        textBoxCustomization.setBorderColor("#FFFFFF");
        textBoxCustomization.setBorderWidth(10);
        textBoxCustomization.setCornerRadius(5);
        textBoxCustomization.setTextColor("#121212");
        textBoxCustomization.setTextFontName("Helvetica");
        textBoxCustomization.setTextFontSize(15);

        ThreeDSecureV2ToolbarCustomization toolbarCustomization =
                new ThreeDSecureV2ToolbarCustomization();
        toolbarCustomization.setBackgroundColor("#FFFFFF");
        toolbarCustomization.setButtonText("Button");
        toolbarCustomization.setHeaderText("Header");
        toolbarCustomization.setTextColor("#121212");
        toolbarCustomization.setTextFontName("Helvetica");
        toolbarCustomization.setTextFontSize(15);

        ThreeDSecureV2UiCustomization customization = new ThreeDSecureV2UiCustomization();
        customization.setButtonCustomization(buttonCustomization);
        customization.setButtonType(ThreeDSecureV2ButtonType.BUTTON_TYPE_CONTINUE);
        customization.setLabelCustomization(labelCustomization);
        customization.setTextBoxCustomization(textBoxCustomization);
        customization.setToolbarCustomization(toolbarCustomization);

        Parcel parcel = Parcel.obtain();
        customization.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        ThreeDSecureV2UiCustomization actual =
                ThreeDSecureV2UiCustomization.CREATOR.createFromParcel(parcel);

        assertEquals("#FF0000", actual.getButtonCustomization().getBackgroundColor());
        assertEquals(5, actual.getButtonCustomization().getCornerRadius());
        assertEquals("#000000", actual.getButtonCustomization().getTextColor());
        assertEquals("Comic Sans", actual.getButtonCustomization().getTextFontName());
        assertEquals(20, actual.getButtonCustomization().getTextFontSize());

        assertEquals("#FF0000",
                actual.getCardinalUiCustomization().getButtonCustomization(ButtonType.CONTINUE)
                        .getBackgroundColor());
        assertEquals(5,
                actual.getCardinalUiCustomization().getButtonCustomization(ButtonType.CONTINUE)
                        .getCornerRadius());
        assertEquals("#000000",
                actual.getCardinalUiCustomization().getButtonCustomization(ButtonType.CONTINUE)
                        .getTextColor());
        assertEquals("Comic Sans",
                actual.getCardinalUiCustomization().getButtonCustomization(ButtonType.CONTINUE)
                        .getTextFontName());
        assertEquals(20,
                actual.getCardinalUiCustomization().getButtonCustomization(ButtonType.CONTINUE)
                        .getTextFontSize());

        assertEquals("#FFFFFF", actual.getLabelCustomization().getHeadingTextColor());
        assertEquals("Times New Roman", actual.getLabelCustomization().getHeadingTextFontName());
        assertEquals(30, actual.getLabelCustomization().getHeadingTextFontSize());
        assertEquals("#121212", actual.getLabelCustomization().getTextColor());
        assertEquals("Helvetica", actual.getLabelCustomization().getTextFontName());
        assertEquals(15, actual.getLabelCustomization().getTextFontSize());

        assertEquals("#FFFFFF",
                actual.getCardinalUiCustomization().getLabelCustomization().getHeadingTextColor());
        assertEquals("Times New Roman", actual.getCardinalUiCustomization().getLabelCustomization()
                .getHeadingTextFontName());
        assertEquals(30, actual.getCardinalUiCustomization().getLabelCustomization()
                .getHeadingTextFontSize());
        assertEquals("#121212",
                actual.getCardinalUiCustomization().getLabelCustomization().getTextColor());
        assertEquals("Helvetica",
                actual.getCardinalUiCustomization().getLabelCustomization().getTextFontName());
        assertEquals(15,
                actual.getCardinalUiCustomization().getLabelCustomization().getTextFontSize());

        assertEquals("#FFFFFF", actual.getTextBoxCustomization().getBorderColor());
        assertEquals(10, actual.getTextBoxCustomization().getBorderWidth());
        assertEquals(5, actual.getTextBoxCustomization().getCornerRadius());
        assertEquals("#121212", actual.getTextBoxCustomization().getTextColor());
        assertEquals("Helvetica", actual.getTextBoxCustomization().getTextFontName());
        assertEquals(15, actual.getTextBoxCustomization().getTextFontSize());

        assertEquals("#FFFFFF",
                actual.getCardinalUiCustomization().getTextBoxCustomization().getBorderColor());
        assertEquals(10,
                actual.getCardinalUiCustomization().getTextBoxCustomization().getBorderWidth());
        assertEquals(5,
                actual.getCardinalUiCustomization().getTextBoxCustomization().getCornerRadius());
        assertEquals("#121212",
                actual.getCardinalUiCustomization().getTextBoxCustomization().getTextColor());
        assertEquals("Helvetica",
                actual.getCardinalUiCustomization().getTextBoxCustomization().getTextFontName());
        assertEquals(15,
                actual.getCardinalUiCustomization().getTextBoxCustomization().getTextFontSize());

        assertEquals("#FFFFFF", actual.getToolbarCustomization().getBackgroundColor());
        assertEquals("Button", actual.getToolbarCustomization().getButtonText());
        assertEquals("Header", actual.getToolbarCustomization().getHeaderText());
        assertEquals("#121212", actual.getToolbarCustomization().getTextColor());
        assertEquals("Helvetica", actual.getToolbarCustomization().getTextFontName());
        assertEquals(15, actual.getToolbarCustomization().getTextFontSize());

        assertEquals("#FFFFFF",
                actual.getCardinalUiCustomization().getToolbarCustomization().getBackgroundColor());
        assertEquals("Button",
                actual.getCardinalUiCustomization().getToolbarCustomization().getButtonText());
        assertEquals("Header",
                actual.getCardinalUiCustomization().getToolbarCustomization().getHeaderText());
        assertEquals("#121212",
                actual.getCardinalUiCustomization().getToolbarCustomization().getTextColor());
        assertEquals("Helvetica",
                actual.getCardinalUiCustomization().getToolbarCustomization().getTextFontName());
        assertEquals(15,
                actual.getCardinalUiCustomization().getToolbarCustomization().getTextFontSize());
    }
}

File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureV2UiCustomizationUnitTest.kt
package com.braintreepayments.api.threedsecure

import android.os.Parcel
import com.cardinalcommerce.shared.models.enums.ButtonType
import kotlinx.parcelize.parcelableCreator
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

@RunWith(RobolectricTestRunner::class)
class ThreeDSecureV2UiCustomizationUnitTest {

    @Test
    fun `sets all cardinal class properties`() {
        val sut = ThreeDSecureV2UiCustomization(
            labelCustomization = ThreeDSecureV2LabelCustomization(),
            buttonCustomization = ThreeDSecureV2ButtonCustomization(),
            buttonType = ThreeDSecureV2ButtonType.BUTTON_TYPE_NEXT,
            textBoxCustomization = ThreeDSecureV2TextBoxCustomization(),
            toolbarCustomization = ThreeDSecureV2ToolbarCustomization()
        )

        val cardinalUiCustomization = sut.cardinalUiCustomization
        assertNotNull(cardinalUiCustomization.labelCustomization)
        assertNotNull(cardinalUiCustomization.getButtonCustomization(ButtonType.NEXT))
        assertNotNull(cardinalUiCustomization.textBoxCustomization)
        assertNotNull(cardinalUiCustomization.toolbarCustomization)
    }

    @Test
    fun `constructor defaults UiCustomization property to empty object`() {
        val sut = ThreeDSecureV2UiCustomization()
        assertNotNull(sut.cardinalUiCustomization)
    }

    @Suppress("LongMethod")
    @Test
    fun `write to parcel`() {
        val buttonCustomization = ThreeDSecureV2ButtonCustomization(
            backgroundColor = "#FF0000",
            cornerRadius = 5,
            textColor = "#000000",
            textFontName = "Comic Sans",
            textFontSize = 20
        )

        val labelCustomization = ThreeDSecureV2LabelCustomization(
            headingTextColor = "#FFFFFF",
            headingTextFontName = "Times New Roman",
            headingTextFontSize = 30,
            textColor = "#121212",
            textFontName = "Helvetica",
            textFontSize = 15
        )

        val textBoxCustomization = ThreeDSecureV2TextBoxCustomization(
            borderColor = "#FFFFFF",
            borderWidth = 10,
            cornerRadius = 5,
            textColor = "#121212",
            textFontName = "Helvetica",
            textFontSize = 15
        )

        val toolbarCustomization = ThreeDSecureV2ToolbarCustomization(
            backgroundColor = "#FFFFFF",
            buttonText = "Button",
            headerText = "Header",
            textColor = "#121212",
            textFontName = "Helvetica",
            textFontSize = 15
        )

        val sut = ThreeDSecureV2UiCustomization(
            buttonCustomization = buttonCustomization,
            buttonType = ThreeDSecureV2ButtonType.BUTTON_TYPE_CONTINUE,
            labelCustomization = labelCustomization,
            textBoxCustomization = textBoxCustomization,
            toolbarCustomization = toolbarCustomization
        )

        val parcel = Parcel.obtain().apply {
            sut.writeToParcel(this, 0)
            setDataPosition(0)
        }
        val parceled = parcelableCreator<ThreeDSecureV2UiCustomization>().createFromParcel(parcel)

        assertEquals("#FF0000", parceled.buttonCustomization?.backgroundColor)
        assertEquals(5, parceled.buttonCustomization?.cornerRadius)
        assertEquals("#000000", parceled.buttonCustomization?.textColor)
        assertEquals("Comic Sans", parceled.buttonCustomization?.textFontName)
        assertEquals(20, parceled.buttonCustomization?.textFontSize)

        assertEquals("#FF0000",
            parceled.cardinalUiCustomization.getButtonCustomization(ButtonType.CONTINUE).backgroundColor)
        assertEquals(5,
            parceled.cardinalUiCustomization.getButtonCustomization(ButtonType.CONTINUE).cornerRadius)
        assertEquals("#000000",
            parceled.cardinalUiCustomization.getButtonCustomization(ButtonType.CONTINUE).textColor)
        assertEquals("Comic Sans",
            parceled.cardinalUiCustomization.getButtonCustomization(ButtonType.CONTINUE).textFontName)
        assertEquals(20,
            parceled.cardinalUiCustomization.getButtonCustomization(ButtonType.CONTINUE).textFontSize)

        assertEquals("#FFFFFF", parceled.labelCustomization?.headingTextColor)
        assertEquals("Times New Roman", parceled.labelCustomization?.headingTextFontName)
        assertEquals(30, parceled.labelCustomization?.headingTextFontSize)
        assertEquals("#121212", parceled.labelCustomization?.textColor)
        assertEquals("Helvetica", parceled.labelCustomization?.textFontName)
        assertEquals(15, parceled.labelCustomization?.textFontSize)

        assertEquals("#FFFFFF",
            parceled.cardinalUiCustomization.labelCustomization.headingTextColor)
        assertEquals("Times New Roman",
            parceled.cardinalUiCustomization.labelCustomization.headingTextFontName)
        assertEquals(30,
            parceled.cardinalUiCustomization.labelCustomization.headingTextFontSize)
        assertEquals("#121212",
            parceled.cardinalUiCustomization.labelCustomization.textColor)
        assertEquals("Helvetica",
            parceled.cardinalUiCustomization.labelCustomization.textFontName)
        assertEquals(15,
            parceled.cardinalUiCustomization.labelCustomization.textFontSize)

        assertEquals("#FFFFFF", parceled.textBoxCustomization?.borderColor)
        assertEquals(10, parceled.textBoxCustomization?.borderWidth)
        assertEquals(5, parceled.textBoxCustomization?.cornerRadius)
        assertEquals("#121212", parceled.textBoxCustomization?.textColor)
        assertEquals("Helvetica", parceled.textBoxCustomization?.textFontName)
        assertEquals(15, parceled.textBoxCustomization?.textFontSize)

        assertEquals("#FFFFFF",
            parceled.cardinalUiCustomization.textBoxCustomization.borderColor)
        assertEquals(10,
            parceled.cardinalUiCustomization.textBoxCustomization.borderWidth)
        assertEquals(5,
            parceled.cardinalUiCustomization.textBoxCustomization.cornerRadius)
        assertEquals("#121212",
            parceled.cardinalUiCustomization.textBoxCustomization.textColor)
        assertEquals("Helvetica",
            parceled.cardinalUiCustomization.textBoxCustomization.textFontName)
        assertEquals(15,
            parceled.cardinalUiCustomization.textBoxCustomization.textFontSize)

        assertEquals("#FFFFFF", parceled.toolbarCustomization?.backgroundColor)
        assertEquals("Button", parceled.toolbarCustomization?.buttonText)
        assertEquals("Header", parceled.toolbarCustomization?.headerText)
        assertEquals("#121212", parceled.toolbarCustomization?.textColor)
        assertEquals("Helvetica", parceled.toolbarCustomization?.textFontName)
        assertEquals(15, parceled.toolbarCustomization?.textFontSize)

        assertEquals("#FFFFFF",
            parceled.cardinalUiCustomization.toolbarCustomization.backgroundColor)
        assertEquals("Button",
            parceled.cardinalUiCustomization.toolbarCustomization.buttonText)
        assertEquals("Header",
            parceled.cardinalUiCustomization.toolbarCustomization.headerText)
        assertEquals("#121212",
            parceled.cardinalUiCustomization.toolbarCustomization.textColor)
        assertEquals("Helvetica",
            parceled.cardinalUiCustomization.toolbarCustomization.textFontName)
        assertEquals(15,
            parceled.cardinalUiCustomization.toolbarCustomization.textFontSize)
    }
}
